ompi/mca/coll/han/coll_han_gatherv.c

```diff
@@ -0,0 +1,381 @@
+    if (han_module->are_ppn_imbalanced) {
+        OPAL_OUTPUT_VERBOSE((30, mca_coll_han_component.han_output,
+                             "han cannot handle gatherv with this communicator (imbalance). Fall "
+                             "back on another component\n"));
+        /* Put back the fallback collective support and call it once. All
+         * future calls will then be automatically redirected.
+         */
+        HAN_LOAD_FALLBACK_COLLECTIVE(han_module, comm, gatherv);
+        return han_module->previous_gatherv(sbuf, scount, sdtype, rbuf, rcounts, displs, rdtype,
+                                            root, comm, han_module->previous_gatherv_module);
+    }
+
+    w_rank = ompi_comm_rank(comm);
+    w_size = ompi_comm_size(comm);
+
+    /* create the subcommunicators */
+    ompi_communicator_t *low_comm
+        = han_module->cached_low_comms[mca_coll_han_component.han_gatherv_low_module];
+    ompi_communicator_t *up_comm
+        = han_module->cached_up_comms[mca_coll_han_component.han_gatherv_up_module];
+
+    /* Get the 'virtual ranks' mapping corresponding to the communicators */
+    vranks = han_module->cached_vranks;
+    /* information about sub-communicators */
+    low_rank = ompi_comm_rank(low_comm);
+    low_size = ompi_comm_size(low_comm);
+    up_rank = ompi_comm_rank(up_comm);
+    up_size = ompi_comm_size(up_comm);
+    /* Get root ranks for low and up comms */
+    mca_coll_han_get_ranks(vranks, root, low_size, &root_low_rank, &root_up_rank);
+
+    OPAL_OUTPUT_VERBOSE((30, mca_coll_han_component.han_output,
+                         "[%d]: Han Gatherv root %d root_low_rank %d root_up_rank %d\n", w_rank,
+                         root, root_low_rank, root_up_rank));
+
+    err = OMPI_SUCCESS;
+    /* #################### Root ########################### */
+    if (root == w_rank) {
+        int need_bounce_buf = 0, total_up_rcounts = 0, *up_displs = NULL, *up_rcounts = NULL,
+            *up_peer_lb = NULL, *up_peer_ub = NULL;
+        char *bounce_buf = NULL;
+
+        low_rcounts = malloc(low_size * sizeof(int));
+        low_displs = malloc(low_size * sizeof(int));
+        if (!low_rcounts || !low_displs) {
+            err = OMPI_ERR_OUT_OF_RESOURCE;
+            goto root_out;
+        }
+
+        int low_peer, up_peer, w_peer;
+        for (w_peer = 0; w_peer < w_size; ++w_peer) {
+            mca_coll_han_get_ranks(vranks, w_peer, low_size, &low_peer, &up_peer);
+            if (root_up_rank != up_peer) {
+                /* Not a local peer */
+                continue;
+            }
+            low_displs[low_peer] = displs[w_peer];
+            low_rcounts[low_peer] = rcounts[w_peer];
+        }
+
+        /* Low Gatherv */
+        low_comm->c_coll->coll_gatherv(sbuf, scount, sdtype, rbuf, low_rcounts, low_displs, rdtype,
+                                       root_low_rank, low_comm,
+                                       low_comm->c_coll->coll_gatherv_module);
```

Comment:
I wonder if this could be overlapped with the global gather. The global gather can use the network for background data movement so once that is started (igatherv) we can take care of the local data movement here (likely boils down to `memcpy`).

Reply:
That's a great idea thank you! Please see update.

For gatherv this didn't help much since the root is waiting for data anyway.

For scatterv this reduced latency significantly because the root is sending data, and the overlap allows it to send to as many peers as possible simultaneously.